---
title: "data_prep"
author: "Ricardo Silva"
date: "02/06/2021"
output: html_document
editor_options: 
  chunk_output_type: console
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      eval=FALSE, 
                      message=FALSE, 
                      warning=FALSE)
```

```{r CLEAR EVERYTHING, eval=FALSE, include=FALSE}
# unload all non-base packages
lapply(names(sessionInfo()$otherPkgs), function(pkgs)
  detach(
    paste0('package:', pkgs),
    character.only = T,
    unload = T,
    force = T
))

rm(list=ls())
```

# Packages and initial steps

```{r load libraries, message=FALSE, warning=FALSE, include=FALSE}
# load the packages
# main packs to start with
pcks <- c('phyloseq', 'microbiome','microbiomeutilities', 'tidyverse')

if(sum(as.numeric(!pcks %in% installed.packages())) != 0){
  installation <- pcks[!pcks %in% installed.packages()]
  for(i in 1:length(installation)) {
    install.packages(installation, dependencies = T)
    break()}
  suppressPackageStartupMessages(
  sapply(pcks,require,character.only = T)
) 
} else {
  suppressPackageStartupMessages(
  sapply(pcks,require,character.only = T)
) 
}

rm(pcks)
```

```{r packages and functions}
# set working directory and seed
setwd("~R/git_hub/MH_2019/data")
folder_path <- "~R/git_hub/MH_2019/output/"

source('~/R/git_hub/MH_2019/script/my_functions.R')
source('~/R/git_hub/MH_2019/script/theme_publication.R')
theme_set(theme_Publication_3())

# set colors
color_distance <- c("D0" = "deepskyblue2", 'D1' = 'brown1',  "D2" = "forestgreen", "D3" = "#440154FF", "D4" = "burlywood")
# color_distance1 <- c('D1' = 'brown1',  "D2" = "forestgreen", "D3" = "#440154FF", "D4" = "burlywood")
color_depth <- c('DP1'="darkorange",'DP2'="purple", "DP3" = "cyan4")
cage_control <- friendly_cols

# check the working directory
current_directory <- getwd()
paste("Working in directory:", current_directory)
```

# Data

## ASV tables

```{r prep data}
# get and prepare the data
# getting the data and filtering

# OTU tables
OTUdf_bac <- read.csv("bac_2019.csv",  row.names = 1, head = TRUE)

# check names
names(OTUdf_bac)
head(rownames(OTUdf_bac))

OTUdf_euk %>% as_tibble()

# taxonomy table
TAXdf_bac <- OTUdf_bac %>% select(Kingdom:Genus)
head(rownames(TAXdf_bac))

# clean ASV tables
OTUdf_bac <- OTUdf_bac %>% select(-seq, -c(Kingdom:Genus))

# metadata table
# meta <- read.csv("meta_env.csv", row.names = 1)
# 
# meta <- meta %>% 
#   unite('dist_depth', Distance, Depth, sep = "_", remove = F) %>% 
#   select(-vol_filtered_ml, -Flag)

# meta with imputed data
# run this chunk to avoid run the others above
meta <- read_csv(paste0(getwd(),"/meta.csv")) %>%
  column_to_rownames("sample_id")

# meta <- new.meta %>% rownames_to_column("sample_id") %>% tibble

meta %>% rownames()

meta %>% pull(dist_depth)

# check if the sample names are the same
# rownames(meta) %in% names(OTUdf_arc)
waldo::compare(sort(rownames(meta)), sort(names(OTUdf_bac)))
```

```{r prep data - tax table}
# check if there is any NA in the taxa table
TAXdf_bac %>% map_df(~sum(is.na(.))) # starts at the class level

# remove NAs in Kingdom and Phylum
# TAXdf <- TAXdf %>% drop_na(Kingdom, Phylum)
# replace NA by unknown
TAXdf_bac <- TAXdf_bac %>% replace(is.na(.), "unknown")
# check again
TAXdf_bac %>% map_df(~sum(is.na(.))) # starts at the class level

# filter contaminants (i.e. mithocondria)
TAXdf_bac %>% map_df(~ sum(str_detect(.,"Mitochondria"))) # 343 ASVs at Family level
TAXdf_bac %>% filter(if_any(everything(), ~str_detect(.,"Mitochondria"))) 
TAXdf_bac <- TAXdf_bac %>% filter(!str_detect(Family, "Mitochondria"))


```

## Picrust table

```{r picrust2 data}
# PICRSUT2 (check picrust2_data_prep.R)
# picrust2 table 
otu_fun_MetaCyc <- read_csv("pathway_abun.csv") %>% 
  rename(pathway_id = pathway) %>%
  mutate(across(where(is.numeric), round))

# picrsut2 "tax table"
tax_fun_MetaCyc <- read_csv("tax_metacyc.csv") 
tax_fun_MetaCyc <- tax_fun_MetaCyc %>% filter(pathway_id %in% otu_fun_MetaCyc$pathway_id)
# check
identical(tax_fun_MetaCyc$pathway_id, otu_fun_MetaCyc$pathway_id)
# as row.names
tax_fun_MetaCyc <- tax_fun_MetaCyc %>% column_to_rownames('pathway_id')
otu_fun_MetaCyc <- otu_fun_MetaCyc %>% column_to_rownames('pathway_id')

# check if the sample names are the same
# rownames(meta) %in% names(otu_fun_MetaCyc)
waldo::compare(sort(rownames(meta)), sort(names(otu_fun_MetaCyc)))
```

```{r prep data - phyloseq objects}

# Bacteria
pseq.bac <- phyloseq(otu_table(as.matrix(as.data.frame(OTUdf_bac)), taxa_are_rows = TRUE),
                     tax_table(as.matrix(TAXdf_bac)),sample_data(meta))

#picrust
pseq.fun <- phyloseq(otu_table(as.matrix(as.data.frame(otu_fun_MetaCyc)), taxa_are_rows = TRUE),
                     tax_table(as.matrix(tax_fun_MetaCyc)), sample_data(meta))

# Saving on object in RData format
save(pseq.bac, file = "pseq.bac.RData")
save(pseq.fun, file = "pseq.fun.RData")

#pseq_df_bac <- microbiomeutilities::phy_to_ldf(pseq.bac[[2]], transform.counts = NULL) # pseq to long format
```


### Filtering and clr transformation

Thus we analyzed this dataset compositionally by first filtering with CodaSeq (min.reads = 2000, min.occurrence = 0.001, min.prop = 0.001), then conducting a center log-ratio transformation (clr), instead of using standard counts and rarefying.

```{r prevalence table}
# Lets generate a prevalence table (number of samples each taxa occurs in) for each taxa.
pseq.bac %>% prevalence_table("Phylum")

 
```

```{r data prep}
# not filter function dataset
# pseq.func <- pseq.bac %>% pluck("function")
# pseq.bac <- pseq.bac %>% list_modify("function" = NULL)

# filtering data and clr-transformation
# min.prop  = minimum proportion in any sample (default)
# min.occurrence = 0.001 # minimum occurrence across all samples (sparsity filter; default)

# function to filter
# pseq.filter <- function(x, min.prop){
#   d.filt0 <- CoDaSeq::codaSeq.filter(otu_table(x), min.reads = 5000, min.prop=0, min.occurrence=0, samples.by.row=FALSE)
#   d.filt <- CoDaSeq::codaSeq.filter(d.filt0,  min.reads = 5000, min.prop = min.prop, max.prop=1, min.occurrence = 0.001, samples.by.row=FALSE)
#   # creatin a TAX table for the phyloseq object
#   TAXdf <- x %>% microbiomeutilities::get_tibble("tax_table") %>% column_to_rownames('column_id')
#   TAX.filt <- base::subset(TAXdf, row.names(TAXdf) %in% row.names(d.filt))
#   pseq.filt <- phyloseq(otu_table(as.matrix(as.data.frame(d.filt)), taxa_are_rows = TRUE), 
#                        tax_table(as.matrix(TAX.filt)), sample_data(as.data.frame(sample_data(x))))
#   return(pseq.filt)
# }

pseq.filter <- function(p){
  # keep taxa seen at least 2 times in at least 2 samples
  p.filt = filter_taxa(p, function(x) sum(x > 2) > 2, TRUE)
#  p.filt1 = prune_taxa(taxa_sums(p.filt) > 10, p.filt) 
return(p.filt)
}

# clr transformation
clr.trans <- function(x){
  pseq <- x
  TAX.filt <- pseq %>% get_tibble("sam_data", column_id = "sample_id") %>% column_to_rownames("sample_id")
  d.filt <- pseq %>% get_tibble("otu_table", column_id = "asv") %>% column_to_rownames("asv")
  d.zero.hand <- t(zCompositions::cmultRepl(t(d.filt), label =0, method="CZM")) # filtered
  d.clr <- CoDaSeq::codaSeq.clr(d.zero.hand, samples.by.row = FALSE) # samples as COLUMN
  df.clr <- phyloseq(otu_table(as.matrix(d.clr), taxa_are_rows = TRUE), 
                     tax_table(as.matrix(tax_table(x))),sample_data(as.data.frame(sample_data(x))))
}

# filtering 
# pseq.bac.filt.0 <- pseq.bac %>% map(.,~pseq.filter(.,min.prop = 0))
# pseq.bac.filt <- pseq.bac %>% map(.,~pseq.filter(.,min.prop = 0.001))
pseq.bac.filt <- pseq.bac %>% pseq.filter
pseq.func.filt <- pseq.fun %>% pseq.filter
# samples removed
# "T1_D1_DP3" and "T1_D2_DP1" for bacteria

# clr transformation
# pseq.bac.0.clr <- pseq.bac.filt.0 %>% map(clr.trans)
pseq.bac.clr <- pseq.bac.filt %>% clr.trans
# picrsut pseq
pseq.func.clr <- pseq.func.filt %>% clr.trans
# ord <- ordinate(pseq.bac.clr[[2]], "RDA", distance = "euclidean")
# plot_ordination(pseq.bac.0.clr[[2]], ord , color = "Depth")

# relative % abundance for bar plot
# pseq.bac.0.rel <- pseq.bac.filt.0 %>% map(.,~transform(.,"compositional"))
pseq.bac.rel <- pseq.bac.filt %>% transform("compositional")

# checking
table(tax_table(pseq.bac)[,"Phylum"])
table(tax_table(pseq.bac.rel)[,"Phylum"])
sample_data(pseq.bac.clr) %>% row.names()

# Saving on object in RData format
# save(pseq.bac.filt.0, file = "pseq.bac.filt.0.RData")
save(pseq.bac.filt , file = "pseq.bac.filt.RData")
save(pseq.func.filt , file = "pseq.func.filt.RData")
# save(pseq.bac.0.clr, file = "pseq.bac.0.clr.RData")
save(pseq.bac.clr, file = "pseq.bac.clr.RData")
save(pseq.func.clr, file = "pseq.func.clr.RData")

pseq.bac.filt %>% map(summarize_phyloseq)
# microbiomeutilities::print_ps()

```

### Checking number of zOTUs before and after filtering

```{r}
# save(pseq.bac.filt.0, file = "pseq.bac.filt.0.RData")
load('pseq.bac.RData')
load('pseq.bac.filt.RData')
load('pseq.fun.RData')
load('pseq.func.filt.RData')
```


```{r}
# total zOTUs
pseq.bac
pseq.bac.filt
pseq.fun
pseq.func.filt

# total reads community 
t.reads.all.f <- tibble(as(sample_data(pseq.bac), "data.frame"),
                 TotalReads = sample_sums(pseq.bac)) %>%
                summarise(sum(TotalReads))

t.reads.filt.f <- tibble(as(sample_data(pseq.bac.filt), "data.frame"),
                 TotalReads = sample_sums(pseq.bac.filt)) %>%
                  summarise(sum(TotalReads))

print(paste(round(((t.reads.all.f - t.reads.filt.f)/t.reads.all.f)*100, 2), "% total reads removed - picrust"))

# total reads function 
t.reads.all.f <- tibble(as(sample_data(pseq.fun), "data.frame"),
                 TotalReads = sample_sums(pseq.fun)) %>%
                summarise(sum(TotalReads))

t.reads.filt.f <- tibble(as(sample_data(pseq.func.filt), "data.frame"),
                 TotalReads = sample_sums(pseq.func.filt)) %>%
                  summarise(sum(TotalReads))

print(paste(round(((t.reads.all.f - t.reads.filt.f)/t.reads.all.f)*100, 2), "% total reads removed - picrust"))




```

```{r}
c("tidymodels", "randomForest", "vip", "ggtern", "factoextra", "stats") %>% purrr::map(., packageVersion) 
c("tidymodels", "randomForest", "vip", "ggtern","factoextra", "stats") %>% purrr::map(., citation) 

```
